include/aikido/statespace/detail/Rn-impl.hpp

```diff
@@ -1,57 +1,275 @@
+#include "aikido/statespace/Rn.hpp"
+
 #include <type_traits>
 
 namespace aikido {
 namespace statespace {
 
+//==============================================================================
+extern template
+class R<0>;
+
+extern template
+class R<1>;
+
+extern template
+class R<2>;
+
+extern template
+class R<3>;
+
+extern template
+class R<6>;
+
+extern template
+class R<Eigen::Dynamic>;
+
+//==============================================================================
 /// \c StateHandle for a \c Rn. The template parameter is
 /// necessary to support both \c const and non-<tt>const</tt> states.
 ///
 /// \tparam _QualifiedState type of \c State being wrapped
 template <class _QualifiedState>
-class RealVectorStateHandle 
-  : public statespace::StateHandle<Rn, _QualifiedState>
+class RStateHandle
+  : public statespace::StateHandle<R<_QualifiedState::DimensionAtCompileTime>, _QualifiedState>
 {
 public:
+  static constexpr int DimensionAtCompileTime = _QualifiedState::DimensionAtCompileTime;
+
+  using Vectord = typename R<DimensionAtCompileTime>::Vectord;
+
   using typename statespace::StateHandle<
-    Rn, _QualifiedState>::State;
+  R<DimensionAtCompileTime>, _QualifiedState>::State;
   using typename statespace::StateHandle<
-    Rn, _QualifiedState>::StateSpace;
+    R<DimensionAtCompileTime>, _QualifiedState>::StateSpace;
   using typename statespace::StateHandle<
-    Rn, _QualifiedState>::QualifiedState;
+    R<DimensionAtCompileTime>, _QualifiedState>::QualifiedState;
 
   using ValueType = std::conditional<std::is_const<QualifiedState>::value,
-    const Eigen::VectorXd, Eigen::VectorXd>;
+    const Vectord, Vectord>;
 
   /// Construct and initialize to \c nullptr.
-  RealVectorStateHandle()
+  RStateHandle()
   {
+    // Do nothing
   }
 
   /// Construct a handle for \c _state in \c _space.
   ///
   /// \param _space state space that created \c _state
   /// \param _state state created by \c _space
-  RealVectorStateHandle(const StateSpace* _space, QualifiedState* _state)
+  RStateHandle(const StateSpace* _space, QualifiedState* _state)
     : statespace::StateHandle<StateSpace, QualifiedState>(_space, _state)
   {
+    // Do nothing
   }
 
   /// Gets the real vector stored in this state.
   ///
   /// \return real vector stored in this state
-  Eigen::Map<const Eigen::VectorXd> getValue() 
+  Eigen::Map<const Vectord> getValue()
   {
     return this->getStateSpace()->getValue(this->getState());
   }
 
   /// Sets the real vector stored in this state.
   ///
   /// \param _value real vector to store in \c _state
-  void setValue(const Eigen::VectorXd& _value)
+  void setValue(const Vectord& _value)
   {
     return this->getStateSpace()->setValue(this->getState(), _value);
   }
 };
 
+//=============================================================================
+template <int N>
+R<N>::R()
+{
+  static_assert(N > -2,
```

Comment:
> N >= 0 || N == Eigen::Dynamic would be cleaner, since it doesn't assume Eigen::Dynamic = -1.

:+1: 



> Can we add a static_assert(N != Eigen::Dynamic) to avoid the edge case below?

I instead changed to throw exceptions for that case because that wouldn't be compiled whenever `N == Eigen::Dynamic`.